src/common/utxobased/keymanager/keymanager.ts

```diff
@@ -287,55 +289,22 @@ function selectedCoinPrefixes(
   prefixes: CoinPrefixes,
   index: number
 ): SelectedCoinPrefixes {
-  const cleanedPrefixes = {
-    messagePrefix:
-      prefixes.messagePrefix[index] ??
-      prefixes.messagePrefix[prefixes.messagePrefix.length - 1],
-    wif: prefixes.wif[index] ?? prefixes.wif[prefixes.wif.length - 1],
-    legacyXPriv:
-      prefixes.legacyXPriv[index] ??
-      prefixes.legacyXPriv[prefixes.legacyXPriv.length - 1],
-    legacyXPub:
-      prefixes.legacyXPub[index] ??
-      prefixes.legacyXPub[prefixes.legacyXPub.length - 1],
-    wrappedSegwitXPriv:
-      prefixes.wrappedSegwitXPriv != null
-        ? prefixes.wrappedSegwitXPriv[index] ??
-          prefixes.wrappedSegwitXPriv[prefixes.wrappedSegwitXPriv.length - 1]
-        : undefined,
-    wrappedSegwitXPub:
-      prefixes.wrappedSegwitXPub != null
-        ? prefixes.wrappedSegwitXPub[index] ??
-          prefixes.wrappedSegwitXPub[prefixes.wrappedSegwitXPub.length - 1]
-        : undefined,
-    segwitXPriv:
-      prefixes.segwitXPriv != null
-        ? prefixes.segwitXPriv[index] ??
-          prefixes.segwitXPriv[prefixes.segwitXPriv.length - 1]
-        : undefined,
-    segwitXPub:
-      prefixes.segwitXPub != null
-        ? prefixes.segwitXPub[index] ??
-          prefixes.segwitXPub[prefixes.segwitXPub.length - 1]
-        : undefined,
-    pubkeyHash:
-      prefixes.pubkeyHash[index] ??
-      prefixes.pubkeyHash[prefixes.pubkeyHash.length - 1],
-    scriptHash:
-      prefixes.scriptHash[index] ??
-      prefixes.scriptHash[prefixes.scriptHash.length - 1],
-    bech32:
-      prefixes.bech32 != null
-        ? prefixes.bech32[index] ?? prefixes.bech32[prefixes.bech32.length - 1]
-        : undefined,
-    // cash addresses are a special case, to accomodate legacy addresses, wrap around to p2pkh
-    cashaddr: prefixes.cashaddr != null ? prefixes.cashaddr[index] : undefined
-  }
-  // skip empty cash addresses
-  if (cleanedPrefixes.cashaddr === '') {
-    cleanedPrefixes.cashaddr = undefined
+  return {
+    messagePrefix: indexAtProtected(prefixes.messagePrefix, index),
+    wif: indexAtProtected(prefixes.wif, index),
+    legacyXPriv: indexAtProtected(prefixes.legacyXPriv, index),
+    legacyXPub: indexAtProtected(prefixes.legacyXPub, index),
+    wrappedSegwitXPriv: indexAtProtected(prefixes.wrappedSegwitXPriv, index),
+    wrappedSegwitXPub: indexAtProtected(prefixes.wrappedSegwitXPub, index),
+    segwitXPriv: indexAtProtected(prefixes.segwitXPriv, index),
+    segwitXPub: indexAtProtected(prefixes.segwitXPub, index),
+    pubkeyHash: indexAtProtected(prefixes.pubkeyHash, index),
+    scriptHash: indexAtProtected(prefixes.scriptHash, index),
+    bech32: indexAtProtected(prefixes.bech32, index),
+    // Special handling for cashaddr to accommodate legacy addresses.
+    // By allowing them to overflow we will fallback to p2pkh.
+    cashaddr: undefinedIfEmptyString((prefixes.cashaddr ?? [])[index])
```

Comment:
Maybe I missed something, but I think this is basically the same:

```js
const selectedCoinPrefixes = (prefixes: CoinPrefixes, index: number): SelectedCoinPrefixes =>
  Object.entries(prefixes).reduce(
    (res, [type, prefixs = []]) => ({
      ...res,
      [type]: type !== 'cashaddr' 
        ? prefixs[Math.min(index, prefixs.length - 1)] 
        : prefixs[index] !== '' 
          ? prefixs[index] 
          : undefined
    }),
    {}
  )
```

Reply:
My original attempt was to use reduce. But I had concerns; which have been resolved with our discussion.

Reply:
This fails test: 

```ts
function selectedCoinPrefixes(
  coinPrefixes: CoinPrefixes,
  index: number
): SelectedCoinPrefixes {
  const hack: SelectedCoinPrefixes = {} as any
  return Object.entries(coinPrefixes).reduce(
    (acc, [prefixType, prefixes]) => ({
      ...acc,
      [prefixType]:
        prefixType === 'cashaddr'
          ? undefinedIfEmptyString((prefixes.cashaddr ?? [])[index])
          : indexAtProtected(prefixes, index)
    }),
    hack
  )
}
```

Unfortunately, I spent too much time on this one so I'm pushing back on this change.